File: app/src/main/java/com/example/domain/usecase/AuthPolicy.kt
package com.example.domain.usecase

import com.example.domain.model.AppUser

/**
 * Pure auth/role decision logic for QueueFuel.
 *
 * Framework-agnostic (no Android/Compose), unit-tested.
 *
 * MVP login is a simple profile registration: name + phone + city, saved
 * directly with `phoneVerified = false`. There is no OTP, no SMS, and no
 * Firebase Phone Auth. Real phone verification can be added server-side later
 * by flipping `phoneVerified` on the user.
 */
object AuthPolicy {

    /** The hardcoded admin phone number for the MVP. */
    const val ADMIN_PHONE: String = "07774564334"

    /** Minimum phone number length required to register. */
    const val MIN_PHONE_LENGTH: Int = 10

    /**
     * True if the phone number (ignoring surrounding whitespace) consists of
     * digits only and meets the minimum length requirement.
     */
    fun isValidPhone(phone: String): Boolean {
        val trimmed = phone.trim()
        return trimmed.length >= MIN_PHONE_LENGTH && trimmed.all { it.isDigit() }
    }

    /**
     * True if the user name is non-blank (required for registration).
     */
    fun isValidName(name: String): Boolean = name.trim().isNotBlank()

    /**
     * True when a registration form (name + phone + selected city) is complete
     * and valid enough to save the user directly.
     */
    fun canRegister(name: String, phone: String, cityId: Int?): Boolean =
        isValidName(name) && isValidPhone(phone) && cityId != null

    /**
     * Resolves the role for a phone number.
     * The admin phone always gets "ADMIN"; everyone else gets "USER".
     */
    fun resolveRole(phone: String): String =
        if (phone == ADMIN_PHONE) "ADMIN" else "USER"

    /**
     * True if the phone belongs to the designated admin.
     */
    fun isAdminPhone(phone: String): Boolean = phone == ADMIN_PHONE

    /**
     * Resolves the effective role once a backend exists: a server-assigned
     * role (`users/{uid}.role` in Firestore, written only by Cloud Functions /
     * the console) always wins; the hardcoded admin phone remains only as the
     * offline-demo fallback when no backend is configured.
     */
    fun effectiveRole(serverRole: String?, phone: String): String =
        serverRole?.takeIf { it.isNotBlank() } ?: resolveRole(phone)

    /**
     * True if the user account is banned and should be rejected at login.
     */
    fun isUserBanned(user: AppUser): Boolean = user.banned
}

File: lib/domain/use_cases/auth_policy.dart
import '../../core/constants/app_constants.dart';
import '../models/app_user.dart';

/// Auth policy ported from AuthPolicy.kt.
/// Handles phone validation, role resolution, and ban checks.
/// Backdoor OTP removed - uses real Firebase Auth exclusively.
class AuthPolicy {
  AuthPolicy._();

  /// The admin phone number
  static const String adminPhone = '07774564334';

  /// Minimum length for a valid phone number
  static const int minPhoneLength = 10;

  /// Validate that a phone number meets minimum length requirements.
  static bool isValidPhone(String phone) {
    final cleaned = phone.replaceAll(RegExp(r'\s+'), '');
    return cleaned.length >= minPhoneLength;
  }

  /// Validate that a user name is non-empty.
  static bool isValidName(String name) {
    return name.trim().isNotEmpty;
  }

  /// Resolve the role for a given phone number.
  /// Returns [UserRole.admin] if the phone matches the admin phone,
  /// otherwise returns [UserRole.user].
  static UserRole resolveRole(String phoneNumber) {
    final cleaned = phoneNumber.replaceAll(RegExp(r'\s+'), '');
    if (isAdminPhone(cleaned)) {
      return UserRole.admin;
    }
    return UserRole.user;
  }

  /// Check if a phone number is the admin phone.
  static bool isAdminPhone(String phoneNumber) {
    final cleaned = phoneNumber.replaceAll(RegExp(r'\s+'), '');
    return cleaned == adminPhone || cleaned.endsWith(adminPhone);
  }

  /// Check if a user is banned.
  static bool isUserBanned(AppUser? user) {
    return user?.banned ?? false;
  }
}
